catlearn/data/utils.py

```diff
@@ -0,0 +1,164 @@
+from typing import Dict, List, Set
+import os
+import warnings
+import argparse
+import logging
+import pickle as pkl
+
+LOGGING_FILE = 'data_utils_debug.log'
+if os.path.exists(os.path.join(os.getcwd(), LOGGING_FILE)):
+    os.remove(os.path.join(os.getcwd(), LOGGING_FILE))
+logging.basicConfig(format='%(asctime)s %(message)s', filename=LOGGING_FILE, level=logging.DEBUG)
+logging.debug('\nStart logging at data.utils on dataset debug level.')
+
+from .config import (raw_dataset_pat, preproc_pat)
+
+RESET_SEQ = "\033[0m"
+COLOR_SEQ = "\033[1;%dm"
+BOLD_SEQ = "\033[1m"
+BLACK, RED, GREEN, YELLOW, BLUE, MAGENTA, CYAN, WHITE = range(8)
+COLORS = {
+    'WARNING': YELLOW,
+    'INFO': WHITE,
+    'DEBUG': BLUE,
+    'CRITICAL': YELLOW,
+    'ERROR': RED
+}
+
+def str_color(color_type: str, string: str) -> str:
+    if color_type == 'W':
+        return COLOR_SEQ % (30 + COLORS['WARNING']) + string + RESET_SEQ
+
+def read_file(path: str) -> List[str]:
+    """Safe reading of file, split it by lines and return in a List"""
+    if not os.path.isfile(path):
+        raise ValueError(f'Dataset path error: ${path}')
+    with open(path, 'r') as f:
+        lines2list = f.read().splitlines()
+    return lines2list
+
+def write_file(path: str, file: Dict):
+    """Writes dictionary to a file
+
+    Args:
+        path (str): [description]
+        file (Dict): [description]
+    """
+    with open(path, 'w') as f:
+        if isinstance(file, dict):
+            for k, v in file.items():
+                f.write(str(k) + preproc_pat['sep'] + str(v) + '\n')
+        elif isinstance(file, set):
+            for v in file:
+                f.write(str(v) + '\n')
+        else:
+            raise ValueError('Unsupported datatype for writing to file.')
+
+def get_entities_relations(ds_list: List[str]) -> [List[str], List[str]]:
+    """Get all entities and relations as 2 separate lists. Used to generate sets of unique values."""
+    triplets = [line.split(raw_dataset_pat['sep']) for line in ds_list]
+    return [h.strip() for h, r, t in triplets] + [t.strip() for h, r, t in triplets], [r.strip() for h, r, t in triplets]
```

Comment:
Could be done as `tuple(zip(*triplets))`, except for the stripping part though - I guess that one can also be treated easily in a previous line.

Whatever the selected code, I tend to favor using `_` for unused variables in list comprehension